embedded/sql/engine.go

```diff
@@ -953,73 +982,71 @@ func EncodeRawValue(val interface{}, colType SQLValueType, asKey bool) ([]byte,
 	return nil, ErrInvalidValue
 }
 
-func EncodeValue(val TypedValue, colType SQLValueType, asKey bool) ([]byte, error) {
+func EncodeAsKey(val interface{}, colType SQLValueType, maxLen int) ([]byte, error) {
 	switch colType {
 	case VarcharType:
 		{
-			strVal, ok := val.(*Varchar)
+			strVal, ok := val.(string)
 			if !ok {
 				return nil, ErrInvalidValue
 			}
 
-			if asKey && len(strVal.val) > len(maxKeyVal(VarcharType)) {
-				return nil, ErrInvalidPK
+			if maxLen > 0 && len(strVal) > maxLen {
```

Comment:
maxLen == 0 when it's not fixed size, e.g. BLOB or VARCHAR types

still work on having max values at sql level is missing, but should be on its own PR

Reply:
There's something I don't get then,

if `maxLen` <= 0 in this code then we should not even try to create the index, right? Shouldn't this be rather:

```suggestion
			if maxLen <= 0 {
			  return nil, ErrInvalidPK
			}
			if len(strVal) > maxLen {
```

Making sure that maxLen is a sane value is critical here because we're building `envc` array few lines below based on that value.

Reply:
maxLen is setting a max length for the column e.g.   varchar[50], so maxLen would be 50

Reply:
Then why we even check if it's > 0?

Reply:
the maxLen is validated when the index is created, but now I placed some validations at the very beginning just for completeness